### PR TITLE
parallelize calibnet check

### PR DESCRIPTION
**Pull request by LesnyRumcajs**
## Summary of changes

Changes introduced in this pull request:

- split calibration check into multiple shell scripts to parallelise the jobs,
- split `Rust` workflow into multiple workflows,
- introduce a test harness to prevent code duplication.

Before it was strictly sequential:
image

The changes intend to make it more parallelisable, with smaller jobs pickable by multiple workers:
image

As a bonus point, every test can be run independently on a local dev machine which should reduce development time and increase the overall quality of dev life.

## Reference issue to close (if applicable)

Closes https://github.com/ChainSafe/forest/issues/2852

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

**Comment by LesnyRumcajs** [comment]
@lemmih if you are happy with the overall architecture of these changes, you will need to adapt the required passing checks as I did a bit of juggling/renaming.

**Comment by lemmih** [comment]
Can we shorten the name of the checks? The `Build and publish Forest Docker image / Build Docker image and push to repositories (pull_request)` is too long to fit on my screen.

File: .github/workflows/forest.yml [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+      - main
+    paths-ignore:
+      - '*.md'
+      - '*.rb'
+  schedule:
+    - cron: "0 0 * * *"
+
+env:
+  CI: 1
+  CARGO_INCREMENTAL: 1
```
Comment (lemmih) [inline .github/workflows/forest.yml:22]: `sccache` cannot cache incremental builds. This value should be `0`.

File: .github/workflows/forest.yml [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        uses: actions/checkout@v3
+      - name: Setup sccache
+        uses: mozilla-actions/sccache-action@v0.0.3
+        timeout-minutes: ${{ fromJSON(env.CACHE_TIMEOUT_MINUTES) }}
+        continue-on-error: true
+      - name: Cargo Check
+        run: cargo check
+        env:
+          CC: "sccache clang"
+          CXX: "sccache clang++"
```
Comment (lemmih) [inline .github/workflows/forest.yml:42]: Should this be moved to the global `env`?

File: .github/workflows/rust-lint.yml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      - main
+    paths-ignore:
+      - '*.md'
+      - '*.rb'
+  schedule:
+    - cron: "0 0 * * *"
+
+env:
+  CI: 1
+  CARGO_INCREMENTAL: 1
```
Comment (lemmih) [inline .github/workflows/rust-lint.yml:22]:
```suggestion
```

File: .github/workflows/unit-tests.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      - main
+    paths-ignore:
+      - '*.md'
+      - '*.rb'
+  schedule:
+    - cron: "0 0 * * *"
+
+env:
+  CI: 1
+  CARGO_INCREMENTAL: 1
```
Comment (lemmih) [inline .github/workflows/unit-tests.yml:25]:
```suggestion
```

**Review by lemmih — APPROVED** [review]

File: .github/workflows/forest.yml (hunk 1 above)
Comment (LesnyRumcajs) [reply .github/workflows/forest.yml:22]: So we had a bug? I mean, the value was there in the `rust.yml`.

Comment (lemmih) [reply .github/workflows/forest.yml:22]: Yeah, it has been a problem for a while.

Comment (lemmih) [reply .github/workflows/forest.yml:22]: Only a slight problem, though. It increases the miss rate of `sccache` by a few percent.

File: .github/workflows/forest.yml (hunk 2 above)
Comment (LesnyRumcajs) [reply .github/workflows/forest.yml:42]: I guess we can try. Hopefully, it won't open any issues, those two variables tend to be *sensitive*.

**Comment by lemmih** [comment]
This is from the Ubuntu build:
```
Cache hits (Rust)                     3
Cache misses (Rust)                 785
```
I do think using S3 would help us get more cache hits.

**Review by elmattic — APPROVED** [review]

**Review by jdjaustin — APPROVED** [review]

**Pull request merged**